### ISSUE TITLE
XRandR changes cause weird Firefox UI update behaviour

### ISSUE THREAD

**Issue by madduck** [awesomeWM/awesome#2762]
Output of `awesome --version`:

awesome v4.3 (Too long)
 • Compiled against Lua 5.3.3 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2

**How to reproduce the issue:**

1. Have Firefox open and on a tag. It does not matter whether the tag is visible or not.
2. Change the XRandR configuration, i.e. adding a screen, or removing one, e.g. by plugging/unplugging an external monitor to a laptop.

**Actual result:**

Firefox no longer receives UI updates unless focus changes, or on some other events. I've recorded [a video to show this behaviour](https://scratch.madduck.net/firefox-xrandr.mp4).

**Expected result:**

Well, just like any other app, including Thunderbird and Chromium, and every other GTK+ or QT app I've tried, this doesn't happen. It's only ever happened with Firefox. Given how EVMH is weird with Firefox anyway, I assume it's specific to Firefox.

I am filing this report here in the hope that Awesome is awesome enough to let me workaround Firefox warts like this, because the developers are simply claiming this doesn't happen anywhere else but with Awesome, and I don't even have to bother filing a Mozilla bug, because that wouldn't get fixed before I die anyway.

**Comment by psychon** [comment]
> I am filing this report here in the hope that Awesome is awesome enough to let me workaround Firefox warts like this, because the developers are simply claiming this doesn't happen anywhere else but with Awesome,

Hm. This is not really helpful, but: Could you test if this happens
- with some other WM (e.g. fluxbox, dwm, ...)
- without any WM at all (just the X11 server and perhaps some terminal to run things)

Another idea would be:
- Run `xwininfo`
- Click Firefox' window
- Note the window id (e.g. `Window id: 0x2000009`)
- Run `xev -id 0x2000009 > log.txt`
- Reproduce the problem
- See if anything useful shows up in the log file (but I do not really expect so)

Note that `xev` is basically more than a key logger, so do not enter any passwords.

**Comment by madduck** [comment]
Here are the XEvents while Firefox is being weirdly rendered, of a simple click on a different tab. You can see the click, as well as the window data update. Nothing about the lack of updating the canvas though:

```
PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x180 (_NET_WM_USER_TIME), time 451811530, state PropertyNewValue

PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x164 (_NET_WM_NAME), time 451811596, state PropertyNewValue

PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x27 (WM_NAME), time 451811597, state PropertyNewValue

PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x19c (_NET_WM_ICON_NAME), time 451811597, state PropertyNewValue

PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x25 (WM_ICON_NAME), time 451811597, state PropertyNewValue

PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x180 (_NET_WM_USER_TIME), time 451813799, state PropertyNewValue

PropertyNotify event, serial 24, synthetic NO, window 0x160002a,
    atom 0x180 (_NET_WM_USER_TIME), time 451813841, state PropertyNewValue
```

So far, I have not been able to reproduce the problem with another window manager, but I'll keep trying.

**Comment by madduck** [comment]
Just an update: I've now tried several times to reproduce this problem with FVWM2 on the same laptop, but have not once seen this behaviour.

And even though it does not happen 100% of the time, it usually doesn't take me more than 2 tries to reproduce the problem when running awesome. So maybe the problem is with Awesome?

**Comment by madduck** [comment]
FTR, I am running with `--no-argb`. Just turned that off (as per suggestion on IRC) and will report back…

**Comment by madduck** [comment]
It happened again, with `--no-argb`. I've now killed `compton` and will report back…

**Comment by psychon** [comment]
Hm. Those are the usual suspect for issues that we manage to trigger in the X11 server. I do not have much experience with debugging GTK and none with Firefox. There is the `GTK_DEBUG=all` environment variable, but I do not even what to look for in the output that it causes.

**Comment by assaron** [comment]
@madduck Did you resolve the issue? I have the same issue, but with xmonad

**Comment by madduck** [comment]
@assaron Nope, still exists. I think it's related to size changes, i.e. if the new screen has the same resolution as the old one, it's fine, but if it's smaller or larger, then FF can't handle it.

**Comment by assaron** [comment]
@madduck It seems I was able to fix the issue by disabling hardware acceleration. Try to run firefox with `--safe-mode` flag, if the issue won't reproduce, then probably that's it.

**Comment by madduck** [comment]
Thanks @assaron for the idea. Unfortunately, this does not get me anywhere. The problem is not reproducible at will. It happens some times when I switch monitor setup, and not at others. So I could be running with `-safe-mode` 100 times and not see the problem once, but on the 101st try it would have happened, except I got bored before. It's like the black swan…

**Comment by assaron** [comment]
@madduck For me it was much more reproducible, so probably we have different issue. Still, I'd recommend just disable hardware acceleration then.

**Comment by madduck** [comment]
@assaron I had disabled hardware accel 2 days ago, and the problem was there this morning, so that unfortunately does not fix it :(

**Comment by ateuber** [comment]
I have the same problem with awesomeVM 4.3 and Firefox (currently 105) maybe once in a month, but it seems to happen randomly without changing XRandR configuration or anything.

**Comment by madduck** [comment]
Conversely, it's not happened to me in a couple of months… I had totally forgotten about this problem.

**Comment by actionless** [comment]
yeah it seems to be fixed on firefox side
